app/src/operations_endpoints.h

```diff
@@ -0,0 +1,402 @@
+    table->put(OperationLog{
+      .status = OperationStatus::Pending,
+      .context = std::move(callback_context),
+    });
+
+    // The AppData allows us to propagate data to the handler. We use it to keep
+    // a reference to the trigger lambda and the bind address. The latter is
+    // needed to build the callback URL, but needs read access to the KV, which
+    // we won't have later on in the commit handler.
+    get_app_data(endpoint_context.rpc_ctx).asynchronous_operation =
+      AsynchronousOperation{
+        .trigger = trigger,
+        .bind_address = get_bind_address(node_context, endpoint_context.tx),
+      };
+  }
+
+  /**
+   * Mark this transaction as having executed a synchronous operation.
+   */
+  void record_synchronous_operation(kv::Tx& tx)
+  {
+    auto operations_table = tx.template rw<OperationsTable>(OPERATIONS_TABLE);
+    operations_table->put(OperationLog{.status = OperationStatus::Registered});
+  }
+
+  /**
+   * Local commit handler for endpoints which create new operations.
+   *
+   * The endpoint handlers don't have access to their transaction ID, and
+   * therefore cannot generate useful responses. This function is therefore
+   * necessary to inform the client of the operation ID.
+   *
+   * This is called for both synchronous and asynchronous operations. In the
+   * case of synchronous operations, the operation's status has already been set
+   * to "registered" in the KV, and once signed, globally committed and
+   * witnessed by the indexing strategy, will be available to the client.
+   *
+   * For asynchronous operations, we must trigger the external process, as
+   * recorded as a lambda in the AppData. Again, we can only do this now that we
+   * know the operation ID, since we need it for the callback URL.
+   *
+   * Note that if the endpoint handlers failed and set a non-2xx HTTP status
+   * code, CCF doesn't apply any of the writes and doesn't call this function.
+   * In this case, it is assumed the handler has set a response body when it set
+   * the erroneous status code.
+   *
+   */
+  void operation_locally_committed_func(
+    ccf::endpoints::CommandEndpointContext& ctx, const ccf::TxID& tx_id)
+  {
+    std::string tx_str = tx_id.to_str();
+    SCITT_INFO("New operation was locally committed with tx={}", tx_str);
+
+    // Even though synchronous operations are set to "Registered" in the KV,
+    // they still need to go through consensus, so we tell the client it is
+    // only pending.
+    GetOperation::Out operation;
+    operation.status = OperationStatus::Pending;
+    operation.operation_id = tx_id;
+
+    ctx.rpc_ctx->set_response_header(http::headers::CCF_TX_ID, tx_str);
+    ctx.rpc_ctx->set_response_header(
+      http::headers::CONTENT_TYPE, http::headervalues::contenttype::JSON);
+    ctx.rpc_ctx->set_response_header(
+      http::headers::LOCATION, fmt::format("/operations/{}", tx_str));
```

Comment:
Location must be an absolute URL.

Reply:
Does it? https://www.rfc-editor.org/rfc/rfc9110.html#section-10.2.2 seems to say relative is fine.

Reply:
Huh, I wasn't aware of that. Looks like the original HTTP 1.1 RFC only allows absolute but the replacement RFC allows both. I haven't seen a relative URL in APIs myself and given that the new RFC just came out last year I'm thinking we should still use absolute URLs for now. We can construct the full URL using the hostname like here: https://github.com/microsoft/CCF/blob/9f7534f0d9af4a95a3ebeb09a1ad62eeed680c22/src/apps/js_generic/js_generic_base.cpp#L189-L200

Reply:
Done.